### ISSUE TITLE
Recurcive transactions

### ISSUE THREAD

**Issue by lamest** [realm/realm-dotnet#1649]
Manually checking everywhere if realm is already in transaction leading to dirty code. There would be nice to use them recursively like the lock construction:

```
using(var transaction=realm.BeginWrite())
{
   realm.Write(()=>{/*some writes*/});
}
```

or 

```
using(var transaction=realm.BeginWrite())
{
   //some writes
   SomeSharedCode(realm);
}
public void SomeSharedCode (Realm realm)
{
   realm.Write(()=>{/*some writes*/});
}
```

**Comment by nirinchev** [comment]
Are you suggesting that `realm.Write` implicitly checks for a transaction and if it exists, simply executes the lambda without trying to open or commit a new one? Or are you suggesting that we add support for nested transactions, where each inner transaction may be committed or rolled back on its own and then the outer transaction controls whether things are ultimately written or not to disk?

**Comment by lamest** [comment]
Me personally would be happy with first option. But nested transactions sounds tasty.

**Comment by nirinchev** [comment]
Sorry for the delayed response. I had a chat with some fellow team members and it seems nested transactions are a bit harder to implement and won't land in the near future. Regarding option 1, I would suggest implementing it as an extension method:

```csharp
public static class RealmExtensions
{
    public static void WriteSafe(this Realm realm, Action action)
    {
        if (realm.IsInTransaction)
        {
            action();
        }
        else
        {
            realm.Write(action);
        }
    }
}
```

**Comment by lengk** [comment]
Is it not possible for a queue to be implemented so that when 
```
Realm.Write(()=>{

})
```
is called an there is another transaction going on, the action is completed when realm is available again?

Also I don't see the method `IsInTransaction` in the realm object as of v2.0.0

**Comment by nirinchev** [comment]
What would be the value of waiting for a transaction to complete just to open a new one? It would be less performant than simply mutating everything in a single transaction.

Re: `IsInTransaction`, it's there: https://realm.io/docs/dotnet/2.1.0/api/reference/Realms.Realm.html#Realms_Realm_IsInTransaction

**Comment by lengk** [comment]
Yes thats true. Is this how it currently works?

If I were to begin a transaction using `Realm.Write()` on **Thread A** and again on **Thread B** immediately after, would the actions be compiled into a single transaction or would **Thread B** receive a `RealmInvalidTransactionException`?

`IsInTransaction` I'm using VSMac and sometimes it doesn't show up. Strange.

**Comment by nirinchev** [comment]
No, realms on separate threads are not affected by what this issue describes. When doing a `Write` on Thread A, you acquire a write lock, which means that until that Write completes, any other thread that calls `Write` will be blocked until the lock is released.
